src/unsafety-checking.md

```diff
@@ -0,0 +1,72 @@
+# Unsafety Checking
+
+Certain expressions in Rust are can violate memory safety and as such need to be
+inside an `unsafe` block or function. The compiler will also warn if an unsafe
+block is used without any corresponding unsafe operations.
+
+## Overview
+
+The unsafety check is located in the [`check_unsafety`] module. It performs a
+walk over the [THIR] of a function and all of its closures and inline constants.
```

Comment:
It would be nice if this explained why we do this in the THIR, but it's not essential.